README
Notes
=====
Adds simple crash management for Trac.

Note: TracCrashDump requires Trac 0.12 or higher.

The following report can be used for crashes group by product and version:

SELECT p.value AS __color__,
c.productname || " " || c.productversion || " (" || c.producttargetversion || ")" AS __group__,
   c.uuid AS _crash,
   c.crashtime as crashtime,
   c.crashusername AS 'Crash user',
   c.crashhostname AS 'Crash hostname',
   c.status,
   c.priority,
   c.component,
   c.version,
   c.milestone,
   c.applicationname as Application,
   c.machinetype || "/" || c.systemname AS 'System name',
   c.osversion AS 'OS Version',
   c.buildtype AS 'Build Type',
   (SELECT group_concat(ticket) from crashdump_ticket where crash=c.id) AS linked_tickets
FROM crashdump c
LEFT JOIN enum p ON p.name = c.priority AND p.type = 'priority'
WHERE c.status <> 'closed'
ORDER BY CAST(p.value AS integer), c.crashtime


debian/changelog
trac-crashdump (0.45) disco; urgency=medium

  * fixed typo

 -- Andreas Roth <aroth@arsoft-online.com>  Thu, 08 Aug 2019 07:56:41 +0200

trac-crashdump (0.44) disco; urgency=medium

  * use UTC timestamps from report

 -- Andreas Roth <aroth@arsoft-online.com>  Thu, 08 Aug 2019 07:27:26 +0200

trac-crashdump (0.43) cosmic; urgency=medium

  * fix issue loading the module list

 -- Andreas Roth <aroth@arsoft-online.com>  Thu, 11 Apr 2019 07:25:31 +0200

trac-crashdump (0.42) cosmic; urgency=medium

  * fix missing variable for search

 -- Andreas Roth <aroth@arsoft-online.com>  Sat, 02 Mar 2019 15:59:10 +0100

trac-crashdump (0.41) bionic; urgency=medium

  * fix issues with missing data

 -- Andreas Roth <aroth@arsoft-online.com>  Thu, 10 May 2018 14:33:11 +0200

trac-crashdump (0.40) bionic; urgency=medium

  * fixed linked_crashes field with empty entries

 -- Andreas Roth <aroth@arsoft-online.com>  Thu, 10 May 2018 13:44:43 +0200

trac-crashdump (0.39) bionic; urgency=medium

  * fixed issue with invalid references in linked_crashes
  * add support for list of system info reports

 -- Andreas Roth <aroth@arsoft-online.com>  Thu, 10 May 2018 12:35:07 +0200

trac-crashdump (0.38) artful; urgency=medium

  * sort memory blocks and regions by base address
  * use mono font for addresses

 -- Andreas Roth <aroth@arsoft-online.com>  Sat, 09 Dec 2017 18:21:30 +0100

trac-crashdump (0.37) artful; urgency=medium

  * fixed missing uuid to submit
  * add crashdump admin page

 -- Andreas Roth <aroth@arsoft-online.com>  Sun, 03 Dec 2017 18:30:48 +0100

trac-crashdump (0.36) artful; urgency=medium

  * add import_crashes script
  * fixed stackdump display

 -- Andreas Roth <aroth@arsoft-online.com>  Sun, 26 Nov 2017 17:43:59 +0100

trac-crashdump (0.35) artful; urgency=medium

  * load crash data from minidump when XML is unavailable

 -- Andreas Roth <aroth@arsoft-online.com>  Wed, 18 Oct 2017 19:12:47 +0200

trac-crashdump (0.34) artful; urgency=medium

  * add system info to crashdumps

 -- Andreas Roth <aroth@arsoft-online.com>  Mon, 16 Oct 2017 07:44:40 +0200

trac-crashdump (0.33) artful; urgency=medium

  * disable tests for now

 -- Andreas Roth <aroth@arsoft-online.com>  Tue, 10 Oct 2017 06:48:57 +0200

trac-crashdump (0.32) artful; urgency=medium

  * a bunch of bug fixes and minor improvements

 -- Andreas Roth <aroth@arsoft-online.com>  Mon, 09 Oct 2017 07:37:12 +0200

trac-crashdump (0.31) yakkety; urgency=medium

  * add option to replace certain owner and reporter name with the real
    usernames

 -- Andreas Roth <aroth@arsoft-online.com>  Mon, 09 Jan 2017 19:05:22 +0100

trac-crashdump (0.30) yakkety; urgency=medium

  * fixed missing uuid in download urls

 -- Andreas Roth <aroth@arsoft-online.com>  Mon, 09 Jan 2017 18:44:38 +0100

trac-crashdump (0.29) yakkety; urgency=medium

  * use proxy object to avoid loading entire XML data
  * allow to download the system info INI file

 -- Andreas Roth <aroth@arsoft-online.com>  Thu, 05 Jan 2017 18:40:54 +0100

trac-crashdump (0.28) yakkety; urgency=medium

  * use jquery to load parts of a crash report on demand
  * improve layout
  * fixed crash submit errors

 -- Andreas Roth <aroth@arsoft-online.com>  Thu, 05 Jan 2017 16:35:38 +0100

trac-crashdump (0.27) yakkety; urgency=medium

  * adjustments for trac 1.2

 -- Andreas Roth <aroth@arsoft-online.com>  Tue, 03 Jan 2017 15:34:20 +0100

trac-crashdump (0.26) xenial; urgency=medium

  * fixed issue when creating new ticket for crash

 -- Andreas Roth <aroth@arsoft-online.com>  Tue, 19 Jul 2016 08:07:56 +0200

trac-crashdump (0.25) xenial; urgency=medium

  * show size of download files in tooltip
  * add report to show the crash uuid

 -- Andreas Roth <aroth@arsoft-online.com>  Fri, 15 Jul 2016 07:24:59 +0200

trac-crashdump (0.24) xenial; urgency=medium

  * fixed wrong dumpdata dir

 -- Andreas Roth <aroth@arsoft-online.com>  Thu, 14 Jul 2016 18:28:25 +0200

trac-crashdump (0.23) xenial; urgency=medium

  * improve error handling and error messages

 -- Andreas Roth <aroth@arsoft-online.com>  Thu, 14 Jul 2016 08:24:33 +0200

trac-crashdump (0.22) xenial; urgency=medium

  * fixed missing SOURCES.txt
  * add server capabilities handler
  * add maximum upload size

 -- Andreas Roth <aroth@arsoft-online.com>  Mon, 11 Jul 2016 06:51:07 +0200

trac-crashdump (0.21) wily; urgency=medium

  * handle xml parser/syntax errors

 -- Andreas Roth <aroth@arsoft-online.com>  Wed, 20 Jan 2016 07:32:42 +0100

trac-crashdump (0.20) wily; urgency=medium

  * fixed win32 syntax errors

 -- Andreas Roth <aroth@arsoft-online.com>  Sun, 06 Dec 2015 18:16:23 +0100

trac-crashdump (0.19) wily; urgency=medium

  * various improvements, see git log

 -- Andreas Roth <aroth@arsoft-online.com>  Sun, 06 Dec 2015 16:41:43 +0100

trac-crashdump (0.18) wily; urgency=medium

  * fixed some syntax errors
  * more CPU info

 -- Andreas Roth <aroth@arsoft-online.com>  Fri, 04 Dec 2015 07:41:49 +0100

trac-crashdump (0.17) wily; urgency=medium

  * add new fields to xml report reader
  * fixed some errors

 -- Andreas Roth <aroth@arsoft-online.com>  Thu, 03 Dec 2015 20:38:06 +0100

trac-crashdump (0.16) wily; urgency=medium

  * improve crash report display
  * fixed some xml parsing mistakes

 -- Andreas Roth <aroth@arsoft-online.com>  Tue, 27 Oct 2015 07:52:57 +0100

trac-crashdump (0.15) wily; urgency=medium

  * improve XML parsing performance

 -- Andreas Roth <aroth@arsoft-online.com>  Mon, 26 Oct 2015 19:35:40 +0100

trac-crashdump (0.14) wily; urgency=medium

  * add request handler to return list of crashes as XML

 -- Andreas Roth <aroth@arsoft-online.com>  Mon, 26 Oct 2015 07:45:49 +0100

trac-crashdump (0.13) wily; urgency=medium

  * fixed handing of linked tickets and crashes

 -- Andreas Roth <aroth@arsoft-online.com>  Sun, 25 Oct 2015 14:36:41 +0100

trac-crashdump (0.12) wily; urgency=medium

  * fixed handling of newly linked tickets to crashes


crashdump/db_default.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# kate: space-indent on; indent-width 4; mixedindent off; indent-mode python;

from trac.db import Table, Column, Index

name = 'crashdump'
schema_version = 13
schema = [
    Table('crashdump', key=('id'))[
        Column('id', type='int', auto_increment=True),
        Column('uuid', size=36),
        Column('type', size=32),
        Column('status'),
        Column('priority', size=32),
        Column('severity', size=32),
        Column('owner'),
        Column('reporter'),
        Column('cc'),
        Column('component'),
        Column('milestone'),
        Column('version'),
        Column('resolution'),
        Column('summary'),
        Column('description'),
        Column('keywords'),
        Column('crashtime', type='int64'),
        Column('reporttime', type='int64'),
        Column('uploadtime', type='int64'),
        Column('changetime', type='int64'),
        Column('closetime', type='int64'),
        Column('applicationname', size=128),
        Column('applicationfile', size=512),
        Column('uploadhostname', size=256),
        Column('uploadusername', size=64),
        Column('crashhostname', size=256),
        Column('crashusername', size=64),
        Column('productname', size=64),
        Column('productcodename', size=64),
        Column('productversion', size=32),
        Column('producttargetversion', size=32),
        Column('buildtype', size=16),
        Column('buildpostfix', size=4),
        Column('machinetype', size=16),
        Column('systemname', size=32),
        Column('osversion', size=32),
        Column('osrelease', size=32),
        Column('osmachine', size=32),

        Column('minidumpfile', size=256),
        Column('minidumpreporttextfile', size=256),
        Column('minidumpreportxmlfile', size=256),
        Column('minidumpreporthtmlfile', size=256),

        Column('coredumpfile', size=256),
        Column('coredumpreporttextfile', size=256),
        Column('coredumpreportxmlfile', size=256),
        Column('coredumpreporthtmlfile', size=256),
        Index(['id'], unique=True),
        Index(['uuid'], unique=True),
    ],
    Table('crashdump_change', key=('crash', 'time', 'field'))[
        Column('crash', type='int'),
        Column('time', type='int64'),
        Column('author', size=256),
        Column('field', size=64),
        Column('oldvalue'),
        Column('newvalue'),
        Index(['crash', 'time', 'field'], unique=True)
    ],
    Table('crashdump_ticket', key=('crash', 'ticket'))[
        Column('crash', type='int'),
        Column('ticket', type='int'),
        Index(['crash', 'ticket'], unique=True)
    ],
    # version 13
    Table('crashdump_stack', key=('crash', 'threadid', 'frameno') )[
        Column('crash', type='int'),
        Column('threadid', type='int'),
        Column('frameno', type='int'),
        Column('module'),
        Column('function'),
        Column('funcoff', type='int'),
        Column('source'),
        Column('line', type='int'),
        Column('lineoff', type='int'),
        Index(['crash', 'frameno']),
        Index(['crash', 'threadid', 'frameno'], unique=True),
    ],
]

def get_current_schema_version(db):
    """Return the current schema version for this plugin."""
    cursor = db.cursor()
    cursor.execute("""
        SELECT value
            FROM system
            WHERE name='%s'
    """ % (name + '_version'))
    row = cursor.fetchone()
    # The expected outcome for any up-to-date installation.
    return row and int(row[0]) or 0

# (table, (column1, column2), ((row1col1, row1col2), (row2col1, row2col2)))
def get_data(db):
    return (
            ('system',
                ('name', 'value'),
                    (
                        (name + '_version', str(schema_version)),
                    )
            ),
            ('report',
                ('title', 'QUERY', 'description'),
                    (
                        ('Active crashes', """
SELECT p.value AS __color__,
""" + db.concat('c.productname'," ",'c.productversion'," (",'c.producttargetversion',")") + """ AS __group__,
   c.uuid AS _crash,
   c.crashtime AS crashtime,
   c.crashusername AS 'Crash user',
   c.crashhostname AS 'Crash hostname',
   c.status,
   c.priority,
   c.component,
   c.version,
   c.milestone,
   c.applicationname AS Application,
   """ + db.concat('c.machinetype',"/",'c.systemname') + """ AS 'System name',
   c.osversion AS 'OS Version',
   c.buildtype AS 'Build Type',
   (SELECT GROUP_CONCAT(ticket) from crashdump_ticket where crash=c.id) AS linked_tickets
FROM crashdump c
LEFT JOIN enum p ON p.name = c.priority AND p.type = 'priority'
WHERE c.status <> 'closed'
ORDER BY """ + db.cast('p.value', 'int') + """, c.crashtime
                        """, """ * List all active crashes by priority.
 * Color each row based on priority."""),
                        ('All crashes', """
SELECT p.value AS __color__,
""" + db.concat('c.productname'," ",'c.productversion'," (",'c.producttargetversion',")") + """ AS __group__,
   c.uuid AS _crash,
   c.crashtime AS crashtime,
   c.crashusername AS 'Crash user',
   c.crashhostname AS 'Crash hostname',
   c.status,
   c.priority,
   c.component,
   c.version,
   c.milestone,
   c.applicationname AS Application,
   """ + db.concat('c.machinetype',"/",'c.systemname') + """ AS 'System name',
   c.osversion AS 'OS Version',
   c.buildtype AS 'Build Type',
   (SELECT GROUP_CONCAT(ticket) from crashdump_ticket where crash=c.id) AS linked_tickets
FROM crashdump c
LEFT JOIN enum p ON p.name = c.priority AND p.type = 'priority'
ORDER BY """ + db.cast('p.value', 'int') + """, c.crashtime
                        """, """ * List all active crashes by priority.
 * Color each row based on priority."""),
                        ('Active crashes with UUID', """
SELECT p.value AS __color__,
""" + db.concat('c.productname'," ",'c.productversion'," (",'c.producttargetversion',")") + """ AS __group__,
   c.uuid AS _crash_uuid,
   c.crashtime AS crashtime,
   c.crashusername AS 'Crash user',
   c.crashhostname AS 'Crash hostname',
   c.status,
   c.priority,
   c.component,
   c.version,
   c.milestone,
   c.applicationname AS Application,
   """ + db.concat('c.machinetype',"/",'c.systemname') + """ AS 'System name',
   c.osversion AS 'OS Version',
   c.buildtype AS 'Build Type',
   (SELECT GROUP_CONCAT(ticket) from crashdump_ticket where crash=c.id) AS linked_tickets
FROM crashdump c
LEFT JOIN enum p ON p.name = c.priority AND p.type = 'priority'
WHERE c.status <> 'closed'
ORDER BY """ + db.cast('p.value', 'int') + """, c.crashtime
                        """, """ * List all active crashes by priority.
 * Color each row based on priority."""),
                        ('All crashes with UUID', """
SELECT p.value AS __color__,
""" + db.concat('c.productname'," ",'c.productversion'," (",'c.producttargetversion',")") + """ AS __group__,
   c.uuid AS _crash_uuid,
   c.crashtime AS crashtime,
   c.crashusername AS 'Crash user',
   c.crashhostname AS 'Crash hostname',
   c.status,
   c.priority,
   c.component,
   c.version,
   c.milestone,
   c.applicationname AS Application,
   """ + db.concat('c.machinetype',"/",'c.systemname') + """ AS 'System name',
   c.osversion AS 'OS Version',
   c.buildtype AS 'Build Type',
   (SELECT GROUP_CONCAT(ticket) from crashdump_ticket where crash=c.id) AS linked_tickets
FROM crashdump c
LEFT JOIN enum p ON p.name = c.priority AND p.type = 'priority'
ORDER BY """ + db.cast('p.value', 'int') + """, c.crashtime
                        """, """ * List all active crashes by priority.
 * Color each row based on priority."""),


('List system information', """

SELECT p.value AS __color__,
""" + db.concat('c.crashhostname'," (",'c.machinetype',"/",'c.systemname',")") + """ AS  __group__,
   c.uuid AS _crash_sysinfo,
   c.crashtime AS crashtime
FROM crashdump c
LEFT JOIN enum p ON p.name = c.priority AND p.type = 'priority'
ORDER BY """ + db.concat('c.crashhostname'," (",'c.machinetype',"/",'c.systemname',")") + """, c.crashtime DESC
                            """, """Lists all system reports grouped by the machine name.

It uses the crash reports in the database to extract this information."""),

            ),
        )
    )
